velox/exec/Unnest.cpp

```diff
@@ -165,100 +166,131 @@ RowVectorPtr Unnest::generateOutput(
 
-    auto nulls =
-        AlignedBuffer::allocate<bool>(numElements, pool(), bits::kNotNull);
-    auto rawNulls = nulls->asMutable<uint64_t>();
+  auto& currentDecoded = unnestDecoded_[channel];
+  auto currentSizes = rawSizes_[channel];
+  auto currentOffsets = rawOffsets_[channel];
+  auto currentIndices = rawIndices_[channel];
 
-    // Make dictionary index for elements column since they may be out of order.
-    index = 0;
-    bool identityMapping = true;
-    for (auto row = start; row < start + size; ++row) {
-      auto maxSize = rawMaxSizes_[row];
+  // Make dictionary index for elements column since they may be out of order.
+  vector_size_t index = 0;
+  bool identityMapping = true;
+  for (auto row = start; row < start + size; ++row) {
+    auto maxSize = rawMaxSizes_[row];
 
-      if (!currentDecoded.isNullAt(row)) {
-        auto offset = currentOffsets[currentIndices[row]];
-        auto unnestSize = currentSizes[currentIndices[row]];
+    if (!currentDecoded.isNullAt(row)) {
+      auto offset = currentOffsets[currentIndices[row]];
+      auto unnestSize = currentSizes[currentIndices[row]];
 
-        if (index != offset || unnestSize < maxSize) {
-          identityMapping = false;
-        }
+      if (index != offset || unnestSize < maxSize) {
+        identityMapping = false;
+      }
 
-        for (auto i = 0; i < unnestSize; i++) {
-          rawElementIndices[index++] = offset + i;
-        }
+      for (auto i = 0; i < unnestSize; i++) {
+        rawElementIndices[index++] = offset + i;
+      }
 
-        for (auto i = unnestSize; i < maxSize; ++i) {
-          bits::setNull(rawNulls, index++, true);
-        }
-      } else if (maxSize > 0) {
-        identityMapping = false;
+      for (auto i = unnestSize; i < maxSize; ++i) {
+        bits::setNull(rawNulls, index++, true);
+      }
+    } else if (maxSize > 0) {
+      identityMapping = false;
 
-        for (auto i = 0; i < maxSize; ++i) {
-          bits::setNull(rawNulls, index++, true);
-        }
+      for (auto i = 0; i < maxSize; ++i) {
+        bits::setNull(rawNulls, index++, true);
       }
     }
+  }
+  return {elementIndices, nulls, identityMapping};
+}
+
+void Unnest::generateOrdinalityVector(
+    vector_size_t start,
+    vector_size_t size,
+    vector_size_t numElements,
+    std::vector<VectorPtr>* outputs) {
+  if (withOrdinality_) {
+    auto ordinalityVector = std::dynamic_pointer_cast<FlatVector<int64_t>>(
+        BaseVector::create(BIGINT(), numElements, pool()));
+
+    // Set the ordinality at each result row to be the index of the element in
+    // the original array (or map) plus one.
+    auto rawOrdinality = ordinalityVector->mutableRawValues();
+    for (auto row = start; row < start + size; ++row) {
+      auto maxSize = rawMaxSizes_[row];
+      std::iota(rawOrdinality, rawOrdinality + maxSize, 1);
+      rawOrdinality += maxSize;
+    }
+
+    // Ordinality column is always at the end.
```

Comment:
Maybe have this method return ordinalityVector and let the caller place it in the output.